### ISSUE TITLE
[rush] Feature request rush scan --json --diff

### ISSUE THREAD

**Issue by chengcyber** [microsoft/rushstack#2450]
## Summary

This is a feature request. For now `rush scan` is targeting to solve phantom dependencies issue. but the stdout log does not help so much to developer.

1. add `--json` flag

`rush scan --json` expects outputing a pure json format (or maybe [NDJSON](https://www.npmjs.com/package/ndjson)) for shell/js script to handle the result directly

current workaround:

```
const { execSync } = require('child_process');
const result = execSync(`rush scan | grep '  ' | tr -d '  '`).toString();

// transform result to a dep name list
const depNames = result.split('\n').filter(Boolean);
```

2. add `--diff` flag

compare the `rush scan` result with current ./package.json 's dep/devDep and list the missing one.

These two flag aims to a further step to solve phantom dependencies issue.

## Other

- [x] I am willing to help implementation

**Comment by iclanton** [comment]
This is a cool idea. It'd be useful to also be able to have the `--diff` flag point out if it appears that you've declared what appears to be an unnecessary dependency.

I'd be happy to accept a contribution here!

**Comment by octogonz** [comment]
As mentioned in the PR notes, I think your `--diff` mode is actually more useful than the current `rush scan` output.

## Current design

When I run your PR branch on the `apps/rush-lib` project in the Rush Stack repo, it shows this output:

**rush scan --diff**
```
Starting "rush scan"

Detected dependencies:
  @azure/identity
  @azure/storage-blob
  @pnpm/link-bins
  @rushstack/heft-config-file
  @rushstack/node-core-library
  @rushstack/package-deps-hash
  @rushstack/rig-package
  @rushstack/stream-collator
  @rushstack/terminal
  @rushstack/ts-command-line
  __
  builtin-modules
  colors
  deploy-metadata
  git-repo-info
  glob
  glob-escape
  ignore
  inquirer
  jszip
  node
  node-fetch
  npm-package-arg
  npm-packlist
  read-package-tree
  resolve
  rx
  semver
  something
  ssri
  strict-uri-encode
  tar
  true-case-path
  wordwrap

Missing dependencies:
+ __
+ deploy-metadata
+ node
+ rx
+ something

Unused dependencies:
- @yarnpkg/lockfile
- cli-table
- https-proxy-agent
- js-yaml
- lodash
- minimatch
- z-schema
```

## Suggested new design

Maybe we could change the design as follows:

**rush scan**
```
Starting "rush scan"

Possible phantom dependencies - these seem to be imported but aren't listed in package.json:
  __
  deploy-metadata  
  node
  rx
  something

Possible unused dependencies - these are listed in package.json but don't seem to be imported:
  @yarnpkg/lockfile
  cli-table
  https-proxy-agent
  js-yaml
  lodash
  minimatch
  z-schema
```

So your `--diff` behavior would be the default.  And instead we could have a parameter `--all` that shows the original report:

**rush scan --all**
```
Starting "rush scan"

Dependencies that seem to be imported by this project:
  @azure/identity
  @azure/storage-blob
  @pnpm/link-bins
  @rushstack/heft-config-file
  @rushstack/node-core-library
  @rushstack/package-deps-hash
  @rushstack/rig-package
  @rushstack/stream-collator
  @rushstack/terminal
  @rushstack/ts-command-line
  __
  builtin-modules
  colors
  deploy-metadata
  git-repo-info
  glob
  glob-escape
  ignore
  inquirer
  jszip
  node
  node-fetch
  npm-package-arg
  npm-packlist
  read-package-tree
  resolve
  rx
  semver
  something
  ssri
  strict-uri-encode
  tar
  true-case-path
  wordwrap
```

Since the `--json` option is intended for tools to consume, seems like we could always include all data.  A tool can simply ignore any unused JSON fields.

**rush scan --json**
```js
{
  "detectedDependencies": [
    "@azure/identity",
    "@azure/storage-blob",
    "@pnpm/link-bins",
    "@rushstack/heft-config-file",
    "@rushstack/node-core-library",
    "@rushstack/package-deps-hash",
    "@rushstack/rig-package",
    "@rushstack/stream-collator",
    "@rushstack/terminal",
    "@rushstack/ts-command-line",
    "__",
    "builtin-modules",
    "colors",
    "deploy-metadata",
    "git-repo-info",
    "glob",
    "glob-escape",
    "ignore",
    "inquirer",
    "jszip",
    "node",
    "node-fetch",
    "npm-package-arg",
    "npm-packlist",
    "read-package-tree",
    "resolve",
    "rx",
    "semver",
    "something",
    "ssri",
    "strict-uri-encode",
    "tar",
    "true-case-path",
    "wordwrap"
  ],
  "missingDependencies": [
    "__",
    "deploy-metadata",
    "node",
    "rx",
    "something"
  ],
  "unusedDependencies": [
    "@yarnpkg/lockfile",
    "cli-table",
    "https-proxy-agent",
    "js-yaml",
    "lodash",
    "minimatch",
    "z-schema"
  ]
}
```
What do you think?

**Comment by octogonz** [comment]
We also could improve the detection.  For example, the RegExp doesn't seem to catch this new pattern used by Rush:
```ts
const cliTable: typeof import('cli-table') = Import.lazy('cli-table', require);
```

And also I noticed it incorrectly matches these pattern, but maybe there isn't a cheap way to fix that:
```ts
 console.log('The link information is read from "deploy-metadata.json" in the same folder.');
```
```ts
      documentation:
        `The Node.js module system allows a project to import NPM packages without explicitly` +
        ` declaring them as dependencies in the package.json file.  Such "phantom dependencies"` +
        ` can cause problems.  Rush and PNPM use symlinks specifically to protect against phantom dependencies.` +
        ` These protections may cause runtime errors for existing projects when they are first migrated into` +
        ` a Rush monorepo.  The "rush scan" command is a handy tool for fixing these errors. It scans the "./src"` +
        ` and "./lib" folders for import syntaxes such as "import __ from '__'", "require('__')",` +
        ` and "System.import('__').  It prints a report of the referenced packages.  This heuristic is` +
        ` not perfect, but it can save a lot of time when migrating projects.`,
      safeForSimultaneousRushProcesses: true,
```

```ts
  // Example: require('something')
```
Probably that should be a separate PR.

**Comment by chengcyber** [comment]
I refactor the PR to apply your new design.

___

The hard part incorrect match. maybe better create a new issue to discussion.

Here is my thoughts:

1. Comment

```
// Example: require('something')
```

we can strip comments before test with RegExp.

2. incorrect `from` match:

> use `"` RegExp here as demo, `'` should be the same

remove:

```
/\bfrom\s*["]([^"]+)["]/
```

add:

```
/**
 * import {
 *   A,
 * } from "foo";
 */
/}\s+from\s*["]([^"]+)["]/,
```

and 

```
/*
 * import A from "foo";
 * import { A } from "foo";
 * import * as A from "foo";
 */
/import[ a-zA-Z{}*]*from\s+["]([^"]+)["]/,
```

and 

```
/*
 * dynamic import
 * import("foo").then(foo => console.log(foo))
 */
/(?<!\.)\bimport\((?:['"].+['"]|`[^$]+`)\)/,
```

3. `const cliTable: typeof import('cli-table') = Import.lazy('cli-table', require);`

hardly handled by RegExp... maybe AST parser comes into rescue.(AST also strips comments lol)

In conclusion, we can add/fix RegExp. but maybe need AST to handle all cases correctly.

For esm, we can use https://github.com/guybedford/es-module-lexer
For cjs, https://github.com/guybedford/cjs-module-lexer or https://www.npmjs.com/package/cjs-exports-parser might be the direction, but they both only detect exports for cjs files.
